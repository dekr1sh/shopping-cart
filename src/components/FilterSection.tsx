import { Card, CardContent, CardFooter, CardHeader, CardTitle } from "@/components/ui/card"
import { Input } from "@/components/ui/input"
import { Label } from "@/components/ui/label"
import { Checkbox } from "@/components/ui/checkbox"
import { Button } from "@/components/ui/button"
import { products } from "@/data/products"
import { getBrands } from "@/lib/productUtils"
import { ChangeEvent } from "react"

interface FilterSectionProps {
  selectedBrands: string[]
  minPrice?: number
  maxPrice?: number
  onBrandChange: (brand: string, checked: boolean) => void
  onMinPriceChange: (value: number | undefined) => void
  onMaxPriceChange: (value: number | undefined) => void
  onClearFilters: () => void
}

export default function FilterSection({
  selectedBrands,
  minPrice,
  maxPrice,
  onBrandChange,
  onMinPriceChange,
  onMaxPriceChange,
  onClearFilters,
}: FilterSectionProps) {
  
  const brands = getBrands(products)

  const handleMinPriceChange = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value ? Number(e.target.value) : undefined
    onMinPriceChange(value)
  }

  const handleMaxPriceChange = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value ? Number(e.target.value) : undefined
    onMaxPriceChange(value)
  }

  return (
    <div>
      <Card className="overflow-hidden">
        <CardHeader className="pb-3">
          <CardTitle className="text-xl text-white">Filters</CardTitle>
        </CardHeader>
        <CardContent className="space-y-8">
          <div>
            <h3 className="font-medium text-lg mb-4 text-white">Brand</h3>
            <div className="space-y-4">
              {brands.map((brand) => (
                <div key={brand} className="flex items-center space-x-3">
                  <Checkbox
                    id={`brand-${brand}`}
                    checked={selectedBrands.includes(brand)}
                    onCheckedChange={(checked) => onBrandChange(brand, checked === true)}
                    className="border-gray-600 cursor-pointer"
                  />
                  <Label
                    htmlFor={`brand-${brand}`}
                    className="text-base font-medium leading-none cursor-pointer text-white"
                  >
                    {brand}
                  </Label>
                </div>
              ))}
            </div>
          </div>

          <div>
            <h3 className="font-medium text-lg mb-4 text-white">Price Range</h3>
            <div className="grid grid-cols-2 gap-3">
              <div className="space-y-2">
                <Label htmlFor="min-price" className="text-sm text-gray-300">
                  Min Price (₹)
                </Label>
                <Input
                  id="min-price"
                  type="number"
                  placeholder="Min"
                  value={minPrice || ""}
                  onChange={handleMinPriceChange}
                  className="bg-[#1a1f2a] border-[#2a3042] text-white"
                />
              </div>
              <div className="space-y-2">
                <Label htmlFor="max-price" className="text-sm text-gray-300">
                  Max Price (₹)
                </Label>
                <Input
                  id="max-price"
                  type="number"
                  placeholder="Max"
                  value={maxPrice || ""}
                  onChange={handleMaxPriceChange}
                  className="bg-[#1a1f2a] border-[#2a3042] text-white"
                />
              </div>
            </div>
          </div>
        </CardContent>
        <CardFooter className="pt-2">
          <Button
            variant="outline"
            className="w-full border-[#2a3042] text-white hover:bg-[#1a1f2a] cursor-pointer"
            onClick={onClearFilters}
          >
            Clear All
          </Button>
        </CardFooter>
      </Card>
    </div>
  )
}

