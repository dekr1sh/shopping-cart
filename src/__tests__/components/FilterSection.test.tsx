import { describe, it, expect, vi } from "vitest";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { screen } from "@testing-library/dom";
import FilterSection from "@/components/FilterSection";

describe("FilterSection", () => {
  const mockHandleBrandChange = vi.fn();
  const mockSetMinPrice = vi.fn();
  const mockSetMaxPrice = vi.fn();
  const mockClearFilters = vi.fn();

  const defaultProps = {
    selectedBrands: [],
    minPrice: undefined,
    maxPrice: undefined,
    onBrandChange: mockHandleBrandChange,
    onMinPriceChange: mockSetMinPrice,
    onMaxPriceChange: mockSetMaxPrice,
    onClearFilters: mockClearFilters,
  };

  it("renders correctly", () => {
    const { container } = render(<FilterSection {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("calls onMinPriceChange when the min price input changes", async () => {
    render(<FilterSection {...defaultProps} />);

    const minPriceInput = screen.getByPlaceholderText("Min");
    await userEvent.type(minPriceInput, "1");

    expect(mockSetMinPrice).toHaveBeenCalledWith(1);
  });

  it("calls onMaxPriceChange when the max price input changes", async () => {
    render(<FilterSection {...defaultProps} />);

    const maxPriceInput = screen.getByPlaceholderText("Max");
    await userEvent.type(maxPriceInput, "5");

    expect(mockSetMaxPrice).toHaveBeenCalledWith(5);
  });
});
