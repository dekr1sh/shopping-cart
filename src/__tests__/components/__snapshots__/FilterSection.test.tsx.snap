// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FilterSection > renders correctly 1`] = `
<div>
  <div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm overflow-hidden"
      data-slot="card"
    >
      <div
        class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 pb-3"
        data-slot="card-header"
      >
        <div
          class="font-semibold text-xl text-white"
          data-slot="card-title"
        >
          Filters
        </div>
      </div>
      <div
        class="px-6 space-y-8"
        data-slot="card-content"
      >
        <div>
          <h3
            class="font-medium text-lg mb-4 text-white"
          >
            Brand
          </h3>
          <div
            class="space-y-4"
          >
            <div
              class="flex items-center space-x-3"
            >
              <button
                aria-checked="false"
                class="peer dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 border-gray-600 cursor-pointer"
                data-slot="checkbox"
                data-state="unchecked"
                id="brand-Nike"
                role="checkbox"
                type="button"
                value="on"
              />
              <label
                class="flex items-center gap-2 select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-base font-medium leading-none cursor-pointer text-white"
                data-slot="label"
                for="brand-Nike"
              >
                Nike
              </label>
            </div>
            <div
              class="flex items-center space-x-3"
            >
              <button
                aria-checked="false"
                class="peer dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 border-gray-600 cursor-pointer"
                data-slot="checkbox"
                data-state="unchecked"
                id="brand-Adidas"
                role="checkbox"
                type="button"
                value="on"
              />
              <label
                class="flex items-center gap-2 select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-base font-medium leading-none cursor-pointer text-white"
                data-slot="label"
                for="brand-Adidas"
              >
                Adidas
              </label>
            </div>
            <div
              class="flex items-center space-x-3"
            >
              <button
                aria-checked="false"
                class="peer dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 border-gray-600 cursor-pointer"
                data-slot="checkbox"
                data-state="unchecked"
                id="brand-New Balance"
                role="checkbox"
                type="button"
                value="on"
              />
              <label
                class="flex items-center gap-2 select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-base font-medium leading-none cursor-pointer text-white"
                data-slot="label"
                for="brand-New Balance"
              >
                New Balance
              </label>
            </div>
            <div
              class="flex items-center space-x-3"
            >
              <button
                aria-checked="false"
                class="peer dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 border-gray-600 cursor-pointer"
                data-slot="checkbox"
                data-state="unchecked"
                id="brand-Puma"
                role="checkbox"
                type="button"
                value="on"
              />
              <label
                class="flex items-center gap-2 select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-base font-medium leading-none cursor-pointer text-white"
                data-slot="label"
                for="brand-Puma"
              >
                Puma
              </label>
            </div>
          </div>
        </div>
        <div>
          <h3
            class="font-medium text-lg mb-4 text-white"
          >
            Price Range
          </h3>
          <div
            class="grid grid-cols-2 gap-3"
          >
            <div
              class="space-y-2"
            >
              <label
                class="flex items-center gap-2 font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-sm text-gray-300"
                data-slot="label"
                for="min-price"
              >
                Min Price (₹)
              </label>
              <input
                class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 flex h-9 w-full min-w-0 rounded-md border px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-[#1a1f2a] border-[#2a3042] text-white"
                data-slot="input"
                id="min-price"
                placeholder="Min"
                type="number"
                value=""
              />
            </div>
            <div
              class="space-y-2"
            >
              <label
                class="flex items-center gap-2 font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 text-sm text-gray-300"
                data-slot="label"
                for="max-price"
              >
                Max Price (₹)
              </label>
              <input
                class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 flex h-9 w-full min-w-0 rounded-md border px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-50 md:text-sm focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive bg-[#1a1f2a] border-[#2a3042] text-white"
                data-slot="input"
                id="max-price"
                placeholder="Max"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex items-center px-6 [.border-t]:pt-6 pt-2"
        data-slot="card-footer"
      >
        <button
          class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs hover:text-accent-foreground dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 px-4 py-2 has-[>svg]:px-3 w-full border-[#2a3042] text-white hover:bg-[#1a1f2a] cursor-pointer"
          data-slot="button"
        >
          Clear All
        </button>
      </div>
    </div>
  </div>
</div>
`;
